src/smart-escrow/SmartEscrow.sol

```diff
@@ -0,0 +1,190 @@
+    event ContractTerminated();
+
+    IERC20 public constant OP_TOKEN = IERC20(0x4200000000000000000000000000000000000042);
+
+    address public beneficiaryOwner;
+    address public beneficiary;
+    uint256 public released;
+    uint256 public immutable start;
+    uint256 public immutable end;
+    uint256 public immutable vestingPeriod;
+    uint256 public immutable initialTokens;
+    uint256 public immutable  vestingEventTokens;
+    bool public contractTerminated;
+
+    /**
+     * @dev Set initial parameters.
+     * @param beneficiaryOwnerAddress Address which can update the beneficiary address.
+     * @param beneficiaryAddress Address which receives tokens that have vested.
+     * @param escrowOwner Address which can terminate the contract.
+     * @param startTimestamp Timestamp of the start of vesting period (or the cliff, if there is one).
+     * @param endTimestamp Timestamp of the end of the vesting period.
+     * @param vestingPeriodSeconds Period of time between each vesting event in seconds.
+     * @param numInitialTokens Number of OP tokens which vest at start time.
+     * @param numVestingEventTokens Number of OP tokens which vest upon each vesting event.
+     */
+    constructor(
+        address beneficiaryOwnerAddress,
+        address beneficiaryAddress,
+        address escrowOwner,
+        uint256 startTimestamp,
+        uint256 endTimestamp,
+        uint256 vestingPeriodSeconds,
+        uint256 numInitialTokens,
+        uint256 numVestingEventTokens
+    ) {
+        if (beneficiaryOwnerAddress == address(0) || beneficiaryAddress == address(0) || escrowOwner == address(0)) {
+            revert AddressIsZeroAddress();
+        }
+        if (startTimestamp == 0) revert StartTimeIsZero();
+        if (startTimestamp > endTimestamp) revert StartTimeAfterEndTime(startTimestamp, endTimestamp);
+        if (vestingPeriodSeconds == 0) revert VestingPeriodIsZeroSeconds();
+
+        beneficiary = beneficiaryAddress;
+        beneficiaryOwner = beneficiaryOwnerAddress;
+        start = startTimestamp;
+        end = endTimestamp;
+        vestingPeriod = vestingPeriodSeconds;
+        initialTokens = numInitialTokens;
+        vestingEventTokens = numVestingEventTokens;
+
+        _transferOwnership(escrowOwner);
+    }
+
+    /**
+     * @dev Allow escrow owner (2-of-2 multisig with beneficiary and benefactor
+     * owners as signers) to terminate the contract.
+     *
+     * Emits a {ContractTerminated} event.
+     */
+    function terminate(address returnAddress) external onlyOwner {
+        contractTerminated = true;
+        emit ContractTerminated();
+        withdrawUnvestedTokens(returnAddress);
+    }
+
+    /**
+     * @dev Allow contract owner to update beneficiary owner address.
+     *
+     * Emits a {BeneficiaryOwnerUpdated} event.
+     */
+    function updateBeneficiaryOwner(address newBeneficiaryOwner) external onlyOwner {
```

Comment:
should we just let them update beneficiary, then? Do we really need beneficiaryOwner, if at the end of the day it's all controlled by owner?

Reply:
It will be nice to have the flexibility to update the owner as needed, without relying on Optimism